### ISSUE TITLE
Method not found error

### ISSUE THREAD

**Issue by drewbj21** [benmccallum/fairybread#30]
Hello,

I am trying to set up fluent validation in a Net5 API using the following packages:

- HotChocolate 11.0.1
- FairyBread 2.0.0-preview.1
- FluentValidation.AspNetCore 9.3.0

Using the above, I am getting the following error:

Method not found: 'System.Threading.Tasks.Task`1<FluentValidation.Results.ValidationResult> FluentValidation.IValidator.ValidateAsync(FluentValidation.ValidationContext, System.Threading.CancellationToken)'.

StackTrace:

   at FairyBread.InputValidationMiddleware.InvokeAsync(IMiddlewareContext context)\r\n   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)\r\n   at FairyBread.InputValidationMiddleware.InvokeAsync(IMiddlewareContext context)\r\n   at lambda_method414(Closure , IMiddlewareContext , InputValidationMiddleware )\r\n   at HotChocolate.Resolvers.FieldClassMiddlewareFactory.<>c__DisplayClass5_0`1.<CreateDelegate>b__1(IMiddlewareContext context)\r\n   at HotChocolate.Execution.Processing.ResolverTask.ExecuteResolverPipelineAsync(CancellationToken cancellationToken)\r\n   at HotChocolate.Execution.Processing.ResolverTask.TryExecuteAsync(CancellationToken cancellationToken)

When I downgrade to FluentValidation.AspNetCore 8.6.3, I am able to validate as expected without this error. Do you know why that is? Let me know if you need anything else from me that could help. Thanks!

**Comment by drewbj21** [comment]
Hey @benmccallum have you been able to look into this?

**Comment by benmccallum** [comment]
Sorry, just replied on slack. Will look probably tomorrow and try to get a proper release our for v11, was taking some time off.

Any repro in a unit test would be greatly appreciated.

**Comment by drewbj21** [comment]
Here's a self contained unit test that reproduces the issue above. It uses the following libraries and is in a net5 project:
- FairyBread v2.0.0-preview.1
- FluentValidation.AspNetCore v9.3.0
- HotChocolate.AspNetCore v11.0.6

```
public class BasicTests
  {
      private static async Task<IRequestExecutor> GetRequestExecutorAsync(Action<IFairyBreadOptions>? configureOptions = null)
      {
          var services = new ServiceCollection();
          services.AddValidatorsFromAssemblyContaining<FooInputDtoValidator>();

          return await services
              .AddGraphQLServer()
              .AddQueryType<QueryType>()
              .AddMutationType<MutationType>()
              .AddFairyBread(options =>
              {
                  options.AssembliesToScanForValidators = new[] { typeof(FooInputDtoValidator).Assembly };
                  configureOptions?.Invoke(options);
              })
              .AddErrorFilter<ValidationErrorFilter>()
              .BuildRequestExecutorAsync();
      }

      [Fact]
      public async Task ErrorRepro()
      {
          // Arrange
          var executor = await GetRequestExecutorAsync();
          var query = "mutation { write(foo: { someInteger: 4, someString: \"test\"}, bar: { emailAddress: \"drew@test.com\" }) }";

          // Act
          var result = await executor.ExecuteAsync(query);

          // Assert
          Assert.True(
              result?.Errors != null 
              && result.Errors.Any() 
              && result.Errors.FirstOrDefault().Message == "Method not found: 'System.Threading.Tasks.Task`1<FluentValidation.Results.ValidationResult> FluentValidation.IValidator.ValidateAsync(FluentValidation.ValidationContext, System.Threading.CancellationToken)'.");

      }

      #region Object set up

      public class QueryType
      {
          public string Read(FooInputDto foo, BarInputDto bar) => $"{foo}; {bar}";
      }

      public class MutationType
      {
          public string Write(FooInputDto foo, BarInputDto bar) => $"{foo}; {bar}";
      }

      public class MyInput : InputObjectType<FooInputDto>
      {
          protected override void Configure(IInputObjectTypeDescriptor<FooInputDto> descriptor)
          {
              descriptor.BindFieldsImplicitly();
          }
      }

      public class FooInputDto
      {
          public int SomeInteger { get; set; }

          public string SomeString { get; set; } = "";

          public override string ToString() =>
              $"SomeInteger: {SomeInteger}, " +
              $"SomeString: {SomeString}";
      }

      public class FooInputDtoValidator : AbstractValidator<FooInputDto>
      {
          public FooInputDtoValidator()
          {
              RuleFor(x => x.SomeInteger).Equal(1);
              RuleFor(x => x.SomeString).Equal("hello");
          }
      }

      public class BarInputDto
      {
          public string EmailAddress { get; set; } = "";

          public override string ToString()
              => $"EmailAddress: {EmailAddress}";
      }

      public class BarInputDtoValidator : AbstractValidator<BarInputDto>
      {
          public BarInputDtoValidator()
          {
              RuleFor(x => x.EmailAddress).NotNull();
          }
      }

      public class BarInputDtoAsyncValidator : AbstractValidator<BarInputDto>
      {
          public BarInputDtoAsyncValidator()
          {
              RuleFor(x => x.EmailAddress)
                  // TODO: Cancellation unit test
                  .MustAsync((val, cancellationToken) => Task.FromResult(val == "ben@lol.com"));
          }
      }

      #endregion
  }
```

**Comment by benmccallum** [comment]
Hmm, it looks like this is unfortunately a breaking change in v9.x of FluentValidation whereby they've removed/changed a method I was using... wonder how I should handle this.

**Comment by benmccallum** [comment]
I believe it's related to this breaking change:
https://docs.fluentvalidation.net/en/latest/upgrading-to-9.html#removal-of-non-generic-validationcontext

As the snippet in your stack trace is referring to the non-generic ValidationContext which is now always generic...
> ```Method not found: 'System.Threading.Tasks.Task`1<FluentValidation.Results.ValidationResult> FluentValidation.IValidator.ValidateAsync(FluentValidation.ValidationContext, System.Threading.CancellationToken)'.```

I've never really had to think about breaking changes on dependencies like this, but I don't believe it's something you can just handle in a similar way to multi-targeting when you can do `#if NETSTANDARD2_0 ... #else ... #endif` for instance. So I think I'm going to have to bump a major version for fairybread, a v3, that also bumps FV.

**Comment by benmccallum** [comment]
I've constrained v2.0.0 of FairyBread to FluentValidation v8+ but < v9, that should make this invalid combination impossible. It's releasing now. 

I'll now create a v3.0.0 of FairyBread that is for FluentValidation >= v9

**Comment by benmccallum** [comment]
Hey @drewbj21, just pushed a `3.0.0-rc.1`, could you give that a go with your codebase and let me know if it's better? Interestingly I didn't see any breaking change on the place I thought it'd be at, someone had already PR'd that back in Sept, so it's quite strange :P 🤞

**Comment by drewbj21** [comment]
Hey @benmccallum, I was able to test with the new version last night and everything looks good! No errors Fluent Validation errors and I'm getting validation errors back as expected.

**Comment by benmccallum** [comment]
Awesome news. 

Fixed in v3.0.0.
